playbooks/fundamentals/template_development_guide.adoc

```diff
@@ -0,0 +1,898 @@
+
+Taking the example `spring-rest` app from the beginning of this guide once again, let's say that we've been experimenting with various tweaks to our application. Since we weren't exactly sure what to do or how, we ended up making some manual changes either using `oc edit` or through the Web Console. We aren't completely sure what changes we made, or how to capture them in the `openjdk-basic-template.yml` file we already have. Exporting our application as a template is a great solution to this problem.
+
+Now, there is some nuance to this method, as not all objects are a good idea to export. `Pod` and `ReplicationController` definitions for example, are intended to be ephemeral, and get generated by the `DeploymentConfig`. Luckily, we can refer back to the set of objects that were originally created during our template exploration.
+
+[source]
+----
+service "spring-rest" configured
+route "spring-rest" configured
+imagestream "spring-rest" configured
+buildconfig "spring-rest" configured
+deploymentconfig "spring-rest" configured
+----
+
+So if we go off of this list, and remembering the `application: spring-rest` label that we placed on those original objects, we should be able to build up our export command.
+
+[source,bash]
+----
+$ oc export bc,is,dc,route,svc -l application=spring-rest --as-template='my-java-app-template'
+apiVersion: v1
+kind: Template
+metadata:
+  creationTimestamp: null
+  name: my-java-app-template
+objects:
+...
+----
+
+This gives us a really solid start to building up an application template. However, this is just the template skeleton and a list of static objects. In order to really make this a reusable templates, we might want to add a few extras, such as:
+
+* Add `parameters` to the template.
+* Further object cleanup. Look for unnecessary fields such as `annotations` and empty `creationTimestamp`s that can be deleted.
+* Make sure we have sensible labeling.
+
+=== Parameter Substitution
+
+Parameters are the means by which we can customize templates. They come in two flavors.
+
+==== String Parameters
+
+String parameters are the most common parameter type. They are represented by single curly braces (e.g. `${FOO}`).
+
+.Example of a String Parameter
+[source,yaml]
+----
+- apiVersion: v1
+  kind: Service
+  metadata:
+    annotations:
+      description: The application's http port.
+    labels:
+      application: ${APPLICATION_NAME}
+    name: ${APPLICATION_NAME}
+  spec:
+    ports:
+    - port: 8080
+      targetPort: 8080
+    selector:
+      deploymentConfig: ${APPLICATION_NAME}
+----
+
+==== Non-String Parameters
+
+Templates also support non-string parameters. They are represented by double curly braces (e.g. `${\{FOO}}`). Non-string parameters provide a way to insert numeric or base64 values into templates.
+
+.Example of a Numeric Non-String Parameter
+[source,yaml]
+----
+spec:
+  ports:
+  - port: ${\{PORT_NUMBER}}
```

Comment:
Variables such as these which have attempted to be escaped are not rendering properly in my asciidoc viewer

Reply:
@sabre1041 yeah, I cannot, for the life of me figure out the right escaping formula for those... i _think_ that this is just an old version of asciidoc that is kind of buggy, but i'm not sure